src/CommonLib/LDAPUtils.cs

```diff
@@ -1014,14 +951,97 @@ public IEnumerable<ISearchResultEntry> QueryLDAP(LDAPQueryOptions options)
             }
         }
 
+        private void HandleSetupException(Exception exception, bool throwException)
+        {
+            _log.LogWarning("Failed to setup LDAP Query Filter: {Message}", exception.Message);
+            if (throwException)
+                throw new LDAPQueryException("Failed to setup LDAP Query Filter", exception);
+        }
+
+        private SearchResponse SendLdapRequest(LdapConnection conn, SearchRequest request, string ldapFilter, ref PageResultResponseControl pageResponse)
+        {
+            _log.LogTrace("Sending LDAP request for {Filter}", ldapFilter);
+            var response = (SearchResponse)conn.SendRequest(request);
+            if (response != null)
+                pageResponse = (PageResultResponseControl)response.Controls.FirstOrDefault(x => x is PageResultResponseControl);
+            return response;
+        }
+
+        private bool HandleServerDownException(string domainName, bool globalCatalog, ref LdapConnection conn, ref int retryCount, ref TimeSpan backoffDelay)
+        {
+            retryCount++;
+
+            // Attempt to acuire a lock
+            if (Monitor.TryEnter(_lockObj))
+            {
+                // If we get the lock, signal our reset event so every other thread waits
+                _connectionResetEvent.Reset();
+                try
+                {
+                    Thread.Sleep(backoffDelay);
+                    // Skip the cache so we don't get an existing connection back
+                    conn = CreateNewConnection(domainName, globalCatalog, skipCache: true).Connection;
+                    if (conn == null)
+                    {
+                        _log.LogError("Unable to create replacement LDAP connection for ServerDown exception. Breaking loop");
+                        return false;
+                    }
+                    _log.LogInformation("Created new LDAP connection after receiving ServerDown from server");
+                }
+                finally
+                {
+                    // Reset our event and release the lock
+                    _connectionResetEvent.Set();
+                    Monitor.Exit(_lockObj);
+                }
+            }
+            else
+            {
+                // If someone else is holding the reset event, we want to just wait and then pull the newly created connection out of the cache
+                // This event will be released after the first entrant thread is done making a new connection
+                // The thread.sleep is to prevent a potential, very unlikely race
+                Thread.Sleep(50);
+                _connectionResetEvent.WaitOne();
+                conn = CreateNewConnection(domainName, globalCatalog).Connection;
+            }
+
+            backoffDelay = GetNextBackoff(retryCount);
+            return true;
+        }
+
+        private void HandleBusyException(ref int retryCount, ref TimeSpan backoffDelay)
```

Comment:
Not the biggest fan of using refs.  Will iterate this and try to remove, but for now I think the current design is simple enough to justify their use here.